### PR TITLE
[aes/dv] adding a test to verify the algorithm matches with the NIST test vectors

### PR DESCRIPTION
**Pull request by rasmus-madsen**
to verify that we actually comply with the NIST standard a test has been made to verify all 3 key lengths.
the test builds on the wakeup test and simply encrypts the plain text from NIST using the key
and then compare to the output from NIST - the inverse process is also checked.

EDIT:
@sriyerg made me aware that we actually have test_vector package under hw/dv/sv
I originally tried to add AES to this package, but I found the similarities are few and the differences many.
So instead of trying to modify the test_ector package to accommodate the AESformat, I created my own package that should together with a python script (not included) that can lift the vectors from the NIST AEStest vector PDFdocument.

the AES test vectors are now in their own class such that CSRNG can leverage them.

### CODE REVIEW

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:29]: Could you use `$onehot` here?

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:42]: Maybe this is obvious if you know the AES block properly, but why does the key need transposing here? I assume that this is something about converting between the format that NIST expects and the one that our AES block expects? Maybe add a note explaining what's going on?

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:45]: What does this wait do? Probably needs explaining (unless it's completely standard in your AES sequences, in which case ignore me!)

Similar for the waits below. Also (super nitty!) maybe it makes sense to group them more consistently. Here, the "wait a bit" is "attached" to the previous operation. Below, it's attached to the spinwait. Maybe shuffle newlines to make them match?

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:48]: Again, maybe a note is needed here about why this needs transposing?

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:20]: Does the NIST document you're following spell it like that? I would have expected "cipher". (Not a problem, but if there's no strong reason for the "y", I'd suggest the more conventional English spelling)

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:65]: Nit: Indentation is off for the continuation line here. Similarly for the next two cases.

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:68]: Should this be `nist_cypher[1] != cypher_text`? (And similarly below)

File: hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv
Comment (rswarbrick) [inline hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv:100]: Maybe this could be `DV_CHECK_FATAL(plain_text == decrypted_text, str)` or something?

**Review by rswarbrick — COMMENTED** [review]
As well as adding a signoff line, could you give a reference to the exact NIST vector in the commit message? (And possibly at the start of `aes_nist_vectors_vseq`). At the moment, it looks like a load of rather arbitrary hex strings! (Obviously, I realise that it isn't, but it's probably helpful to include a pointer).

**Comment by rasmus-madsen** [comment]
> As well as adding a signoff line, could you give a reference to the exact NIST vector in the commit message? (And possibly at the start of `aes_nist_vectors_vseq`). At the moment, it looks like a load of rather arbitrary hex strings! (Obviously, I realise that it isn't, but it's probably helpful to include a pointer).

@rswarbrick I completely refactored the code - and the vectors so this is basically a new commit. I am sorry for the double work but I would very much appreciate it if you could re-review

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (sriyerg) [inline hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @weicaiyang / @cindychip 
Don't we have RAL APIs to generate this condition?

**Comment by sriyerg** [comment]
> have test_vector package under hw/dv/sv
> I originally tried to add AES to this package, but I found the similarities are few and the differences many.
> So instead of trying to modify the test_ector package to accommodate the AESformat, I created my own package that should together with a python script (not included) that can lift the vectors from the NIST AEStest vector PDFdocument.
> 
> the AES test vectors are now in their own class such that CSRNG can leverage them.

I would probably urge you to integrate with the `test_vectors_pkg` so that its all in one place and we can reuse at the chip level. We can discuss this in the next meeting?

**Comment by rasmus-madsen** [comment]
> > have test_vector package under hw/dv/sv
> > I originally tried to add AES to this package, but I found the similarities are few and the differences many.
> > So instead of trying to modify the test_ector package to accommodate the AESformat, I created my own package that should together with a python script (not included) that can lift the vectors from the NIST AEStest vector PDFdocument.
> > the AES test vectors are now in their own class such that CSRNG can leverage them.
> 
> I would probably urge you to integrate with the `test_vectors_pkg` so that it's all in one place and we can reuse it at the chip level. We can discuss this in the next meeting?

@sriyerg sure we can discuss it next meeting.
but I think someone - will need to overhaul the existing test_vector package to make it compatible with the AES vectors.
so this might very well boil down to the question of anyone has time to do this.

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (cindychip) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: Hi Sri, thanks for the comment.
This is what we have done for shadow_reg test:
https://github.com/lowRISC/opentitan/blob/master/hw/dv/sv/cip_lib/seq_lib/cip_base_vseq.sv#L792
We are using `csr_poke` method. The CSR backdoor path is implemented here: https://github.com/lowRISC/opentitan/blob/master/hw/ip/aes/data/aes.hjson#L461

For your case, I think you can try:
csr_poke(.ptr(ral.ctrl_shadowed), .value(mal_error), .kind(BkdrRegPathRtlCommitted), .predict(1));

File: hw/ip/aes/dv/test_vectors/nist_vectors.sv
Comment (weicaiyang) [inline hw/ip/aes/dv/test_vectors/nist_vectors.sv:28]: nit: 2 space indent

File: hw/ip/aes/dv/test_vectors/nist_vectors.sv
Comment (weicaiyang) [inline hw/ip/aes/dv/test_vectors/nist_vectors.sv:47]: perhaps move the class to a separate file

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @sriyerg how to do this was discussed previously https://github.com/lowRISC/opentitan/issues/5652
unfortunately, I can't seem to find the notes from the dv meeting where we discussed this.

anyway, this follows the method used by the keymgr
only difference is that I use uvm_hdl_force vs force. (uvm_hdl_force searches for the path at run times which allows me to use it in the sequence)

another point to make is the RAL requires a register, which is valid for this individual case - but I am adding 20+ signals that are not all registers that will require a force. and I prefer to keep them all in the same format.

do you have any objections to using force?

**Comment by rasmus-madsen** [comment]
FYI I am fixing the lint stuff. 
it is all in the auto gen file from my python script

File: hw/ip/aes/dv/test_vectors/nist_vectors.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/test_vectors/nist_vectors.sv:47]: the file is autogenerated from a pdf.
I should probably add some text in the top describing this.
but I prefer to not split it.

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: Since this is a CSR, we could use RAL API to do backdoor force. And you don't need to worry about the hierarchy path, as CSR tests have ensured that it's correct.

BTW, looks like you want to check when the value isn't one-hot, it should trigger error / alert. But you just change the committed value, which actually triggers shadowed reg error due to mismatch between committed value and shadowed value.

perhaps we should force both values
```
csr_poke(.ptr(ral.ctrl_shadowed), .value(mal_error), .kind(BkdrRegPathRtlCommitted));
csr_poke(.ptr(ral.ctrl_shadowed), .value(mal_error), .kind(BkdrRegPathRtlShadow));
```

Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: For uvm_hdl_force vs force, I think either way is fine. We use both in our env.

Comment (sriyerg) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: The RAL APIs already factor in the paths, and is hence preferred (applies to all registers). We have backdoor reads and writes to registers working, so there is no need to use hardcoded paths separately. 

For other non-register signals that you may want to force / probe, why not use an SV interface? Bind the said interface to AES hierarchy in the tb and use partial paths into the design - you dont need to stringify anything - you can use SV forces which is simpler than `uvm_hdl_force`.

File: hw/ip/aes/dv/test_vectors/nist_vectors.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/test_vectors/nist_vectors.sv]: fixed

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @sriyerg @weicaiyang 
the force is anyway not a part of this commit (only the clean up of warnings.)
Let's push the discussion for the next commit that adds the other signals. to not stall this one

Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: > BTW, looks like you want to check when the value isn't one-hot, it should trigger error / alert. But you just change the committed value, which actually triggers shadowed reg error due to mismatch between committed value and shadowed value.

Maybe I am misunderstanding what you are trying to say.
but the alert is generated by the DUT and not the shadow reg.

**Comment by sriyerg** [comment]
> > > have test_vector package under hw/dv/sv
> > > I originally tried to add AES to this package, but I found the similarities are few and the differences many.
> > > So instead of trying to modify the test_ector package to accommodate the AESformat, I created my own package that should together with a python script (not included) that can lift the vectors from the NIST AEStest vector PDFdocument.
> > > the AES test vectors are now in their own class such that CSRNG can leverage them.
> > 
> > 
> > I would probably urge you to integrate with the `test_vectors_pkg` so that it's all in one place and we can reuse it at the chip level. We can discuss this in the next meeting?
> 
> @sriyerg sure we can discuss it next meeting.
> but I think someone - will need to overhaul the existing test_vector package to make it compatible with the AES vectors.
> so this might very well boil down to the question of anyone has time to do this.

I looked at the `test_vectors_pkg` implementation - it looked fairly straightforward to me to get it updated for AES case. You would check in the vectors [here](https://github.com/lowRISC/opentitan/tree/master/hw/dv/sv/test_vectors/vectors) and add data structure for each type of AES (where type is a combination of length and mode). The parsing structure needs a little bit of massaging to get it working, but it does not seem too bad.

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (sriyerg) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: It's a shame that whoever reviewed the last PR when you added these lines, did not point out that not only we have methods to generate this force, but also the entire hierarchy to the CSR is being hardcoded here, which is a bad idea in terms of reuse. 

I am sorry if you feel like this PR is getting stalled, but the point of reviews is to spot these things. Why not just replace those lines with `csr_poke` as suggested by Cindy and Weicai? IMO its a very simple ask.

**Comment by sriyerg** [comment]
> > > > have test_vector package under hw/dv/sv
> > > > I originally tried to add AES to this package, but I found the similarities are few and the differences many.
> > > > So instead of trying to modify the test_ector package to accommodate the AESformat, I created my own package that should together with a python script (not included) that can lift the vectors from the NIST AEStest vector PDFdocument.
> > > > the AES test vectors are now in their own class such that CSRNG can leverage them.
> > > 
> > > 
> > > I would probably urge you to integrate with the `test_vectors_pkg` so that it's all in one place and we can reuse it at the chip level. We can discuss this in the next meeting?
> > 
> > 
> > @sriyerg sure we can discuss it next meeting.
> > but I think someone - will need to overhaul the existing test_vector package to make it compatible with the AES vectors.
> > so this might very well boil down to the question of anyone has time to do this.
> 
> I looked at the `test_vectors_pkg` implementation - it looked fairly straightforward to me to get it updated for AES case. You would check in the vectors [here](https://github.com/lowRISC/opentitan/tree/master/hw/dv/sv/test_vectors/vectors) and add file path arrays for each type of AES (where type is a combination of length and mode). The parsing structure needs a little bit of massaging to get it working, but it does not seem too bad.

I have filed issue #7143 to track this. Please add `// TODO(#7143): ` comment in the code to reference this - we don't need to solve it right away.

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: > Maybe I am misunderstanding what you are trying to say.
but the alert is generated by the DUT and not the shadow reg.

The reg you force here is a shadow reg, so if you just force the committed value, then it will mismatch with the shadowed value, which triggers an alert. So I think the alert you see here may not be the one you want.

File: hw/ip/aes/dv/test_vectors/read_nist_vectors.py
Comment (udinator) [inline hw/ip/aes/dv/test_vectors/read_nist_vectors.py:1]: nit: need to add license header

File: hw/ip/aes/dv/test_vectors/read_nist_vectors.py
Comment (udinator) [inline hw/ip/aes/dv/test_vectors/read_nist_vectors.py:45]: nit: `sript` -> `script`

File: hw/ip/aes/dv/test_vectors/read_nist_vectors.py
Comment (udinator) [inline hw/ip/aes/dv/test_vectors/read_nist_vectors.py:24]: is this `nist_vectors.txt` the downloaded (and renamed) vector file from NIST website?

Comment (rasmus-madsen) [reply hw/ip/aes/dv/test_vectors/read_nist_vectors.py:24]: no it is a converted pdf, you can download it from the link in the file (also mentioned in the README)
do you want me to add the txt file also? might be helpful - but I obviously can add a license header to that file. 
so CI would probably fail

Comment (udinator) [reply hw/ip/aes/dv/test_vectors/read_nist_vectors.py:24]: No need to add that file as of now, it's fine to accept the "hardcoded" `nist_vectors_pkg.sv` that you've added as we've essentially hardcoded all of the other NIST vectors.
We'll take care of this at a later time.

**Comment by rasmus-madsen** [comment]
@udinator I removed the python script again as it did not pass linting.
I will add it in a draft or after this goes in

**Comment by rasmus-madsen** [comment]
@sriyerg @weicaiyang I think all concerns have been addressed.
any more comments?

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @rasmus-madsen any update on this?

Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: The signal that is forced is select by @vogelpi 
I double-checked it does cause a fatal as expected. if it is a shadow_reg alert I would expect it to be recoverable.

Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: It triggers storage error in this shadow_reg as we backdoor change `committed_q`. And it is a fatal error.

>   assign err_storage = (~shadow_q != committed_q);

perhaps @vogelpi can confirm if this path is right? It should be a place that is forced to a non-one-hot value to trigger the error. If it's just to trigger a storage error for shadow reg, the common test added by Cindy can cover that.

Comment (rasmus-madsen) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @weicaiyang I agree this is an important detail to investigate
but it is not a part of this PR which is only about the nist vectors.

the reason this line even shows up is because I lint cleaned it!

Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: since we found 2 potential issue/improvement here (1. can consider to use csr_poke as the hierarchy path is well maintained in the generated RAL model. 2. this may end up triggering a storage error of shadowed reg), can you add a TODO with comments? Then we can merge this PR and address it later.

**Comment by rasmus-madsen** [comment]
@weicaiyang @cindychip @sriyerg @rswarbrick 
please comment or approve.

File: hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv
Comment (weicaiyang) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: @rasmus-madsen can you address my last comment to add a TODO? Thanks

Comment (vogelpi) [reply hw/ip/aes/dv/env/seq_lib/aes_alert_reset_vseq.sv:39]: Sorry for not responding earlier and blocking this PR. I accidentally overlooked your request for my input.

I can confirm that the signal used by @rasmus-madsen is correct. I've compiled a list with all signals to check and that signal is on the list. So from my side, you have green light.

File: hw/ip/aes/dv/test_vectors/nist_vectors_pkg.sv
Comment (vogelpi) [inline hw/ip/aes/dv/test_vectors/nist_vectors_pkg.sv:35]: I strongly recommend to both print a note here and include comments in this file informing the user about endianness as it this keeps confusing people over and over again. For the data in the NIST document we have the following value e.g. for 128-bit ECB key:
```
2B7E1516 28AED2A6 ABF71588 09CF4F3C
``` 
the real key used is then:
```
byte 0: 2B
byte 1: 7E
...
byte 16: 3C
```
Unfortunately, this is not explicitly stated in the linked document, but it is stated in the [NIST FIPS-197 standard (AES)](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.197.pdf): See Section 3.3.

Below you have the following value:
```
nist_vectors[0].key  = 256'h2B7E151628AED2A6ABF7158809CF4F3C00000000000000000000000000000000
```
and then you convert the endianness at run time in `hw/ip/aes/dv/env/seq_lib/aes_nist_vectors_vseq.sv`. I don't want to argue which way makes more sense for the pgk file, but it should be absolutely clear for users what's what. Otherwise people will get confused and start to look for potential bugs in the wrong place.

**Review by vogelpi — APPROVED** [review]
Thanks @rasmus-madsen for this PR and sorry for having blocked it for so long.

The PR looks good to me but I would appreciate if you could include a note + printf about the used data format (see separate comment). So far, whenever someone new has started using our AES IP (RTL or DV) I had to go over this again. I've already documented the format in the RTL and it certainly wouldn't hurt here as well.

**Review by weicaiyang — APPROVED** [review]
LGTM, thanks

**Pull request merged**
